src/private_broadcast.cpp

```diff
@@ -0,0 +1,149 @@
+            most_urgent = tx;
+            most_urgent_prio = tx_prio;
+            most_urgent_it = it;
+        }
+    }
+
+    if (most_urgent.has_value()) {
+        auto& sent_to{most_urgent_it->second};
+        sent_to.push_back(SendStatus{.nodeid = will_send_to_nodeid,
+                                     .picked = NodeClock::now(),
+                                     .confirmed = std::nullopt});
+    }
+
+    return most_urgent;
+}
+
+std::optional<CTransactionRef> PrivateBroadcast::GetTxForNode(const NodeId& nodeid)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto tx_and_status{GetSendStatusByNode(nodeid)};
+    if (tx_and_status.has_value()) {
+        return tx_and_status.value().tx;
+    }
+    return std::nullopt;
+}
+
+void PrivateBroadcast::NodeConfirmedReception(const NodeId& nodeid)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto tx_and_status{GetSendStatusByNode(nodeid)};
+    if (tx_and_status.has_value()) {
+        tx_and_status.value().send_status.confirmed = NodeClock::now();
+    }
+}
+
+bool PrivateBroadcast::DidNodeConfirmReception(const NodeId& nodeid)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto tx_and_status{GetSendStatusByNode(nodeid)};
+    if (tx_and_status.has_value()) {
+        return tx_and_status.value().send_status.confirmed.has_value();
+    }
+    return false;
+}
+
+std::vector<CTransactionRef> PrivateBroadcast::GetStale() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+{
+    LOCK(m_mutex);
+    const auto stale_time{NodeClock::now() - STALE_DURATION};
+    std::vector<CTransactionRef> stale;
+    for (const auto& [tx, send_status] : m_transactions) {
+        const Priority p{DerivePriority(send_status)};
+        if (p.last_confirmed < stale_time) {
+            stale.push_back(tx);
+        }
+    }
+    return stale;
+}
+
+std::strong_ordering PrivateBroadcast::Priority::operator<=>(const Priority& other) const
+{
+    const auto this_tie{std::tie(num_picked, num_confirmed, last_picked, last_confirmed)};
+    const auto other_tie{std::tie(other.num_picked, other.num_confirmed, other.last_picked, other.last_confirmed)};
+
+    if (this_tie < other_tie) { // Smaller num_picked, num_confirmed or older times mean greater priority.
+        return std::strong_ordering::greater;
+    }
+    if (this_tie == other_tie) {
+        return std::strong_ordering::equivalent;
+    }
+    return std::strong_ordering::less;
+}
+
+PrivateBroadcast::Priority PrivateBroadcast::DerivePriority(const std::vector<SendStatus>& sent_to)
+{
+    Priority p;
+    p.num_picked = sent_to.size();
+    for (const auto& send_status : sent_to) {
+        p.last_picked = std::max(p.last_picked, send_status.picked);
+        if (send_status.confirmed.has_value()) {
+            ++p.num_confirmed;
+            p.last_confirmed = std::max(p.last_confirmed, send_status.confirmed.value());
```

Comment:
we have already validated `if (send_status.confirmed) {` above, we could reduce the noise a bit:

```suggestion
            p.last_confirmed = std::max(p.last_confirmed, *send_status.confirmed);
```

Reply:
Both `*foo` and `foo.value()` are valid. I find the latter more readable and prefer to keep it. Same with `if (foo)` vs `if (foo.has_value())`.